### PR TITLE
Fix flaky 3PID inhibit error unit tests

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes flakes in the following unit tests:

- `tests.rest.client.test_account.PasswordResetTestCase.test_password_reset_bad_email_inhibit_error`
- `tests.rest.client.test_register.RegisterRestServletTestCase.test_request_token_existing_email_inhibit_error`

These tests could flake as the endpoints would intentionally wait for 100-1000ms in order to prevent a timing attack allowing one to deduce whether an email was already registered.

The tests were only waiting 1000ms for the endpoint to return, so a randomly-chosen jitter any higher than ~900ms would cause the test to fail.

---

I'm unsure why this started failing recently, as the jitter has been in place for years. There was [a recent change](https://github.com/element-hq/synapse/pull/19229) with the `Duration` work - we used `randrandom.randint(1, 10) / 10` and now `Duration(milliseconds=random.randint(100, 1000))`, but the math is technically equivalent.

### Dev notes

There are multiple places we add jitter in both `synapse/rest/client/account.py` and `synapse/rest/client/register.py`

https://github.com/element-hq/synapse/blob/963e486a70fa841f89dc7c84d551e29e3335937d/synapse/rest/client/account.py#L125-L131

https://github.com/element-hq/synapse/blob/963e486a70fa841f89dc7c84d551e29e3335937d/synapse/rest/client/register.py#L157-L165

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Review by reivilibre — APPROVED** [review]

File: tests/rest/client/test_account.py [hunk 1]
```diff
@@ -325,7 +325,13 @@ def test_password_reset_bad_email_inhibit_error(self) -> None:
         email = "test@example.com"
 
         client_secret = "foobar"
-        session_id = self._request_token(email, client_secret)
+        session_id = self._request_token(
+            email,
+            client_secret,
+            # The endpoint intentionally adds up to 1000ms of jitter to avoid
+            # leaking whether the email address is bound to an account.
+            timeout_ms=3000,
```
Comment (MadLittleMods) [inline tests/rest/client/test_account.py:333]: > The tests were only waiting 1000ms for the endpoint to return, so a randomly-chosen jitter any higher than ~900ms would cause the test to fail.

I'm not understanding the logic behind 3000ms?

File: tests/rest/client/test_account.py [hunk 2]
```diff
@@ -364,6 +370,7 @@ def _request_token(
         client_secret: str,
         ip: str = "127.0.0.1",
         next_link: str | None = None,
+        timeout_ms: int = 1000,
```
Comment (MadLittleMods) [inline tests/rest/client/test_account.py:373]: We can use the `Duration` class

Comment (MadLittleMods) [inline tests/rest/client/test_account.py:373]: Why `1000` default if that doesn't work in these cases?

File: tests/server.py [hunk 3]
```diff
@@ -371,6 +371,7 @@ def make_request(
     await_result: bool = True,
     custom_headers: Iterable[CustomHeaderType] | None = None,
     client_ip: str = "127.0.0.1",
+    timeout_ms: int = 1000,
```
Comment (MadLittleMods) [inline tests/server.py:374]: Historically, we instead used this pattern:

```python
channel = self.make_request(
    "GET",
    f"/_matrix/media/r0/thumbnail/{self.media_id}{params}",
    await_result=False,
)
channel.await_result(timeout_ms=1000)
```

The new shortcut doesn't seem bad though.

File: tests/rest/client/test_account.py (hunk 1 above)
Comment (MadLittleMods) [reply tests/rest/client/test_account.py:333]: This should actually be fine with 1000ms (the default).

The only reason it's not is because [`await_result(...)`](https://github.com/element-hq/synapse/blob/963e486a70fa841f89dc7c84d551e29e3335937d/tests/server.py#L301-L312) is a bit flaky. For example, the first `advance(0.1)` in `await_result(...)` triggers any scheduled work (like database [`ThreadPool`](https://github.com/element-hq/synapse/blob/963e486a70fa841f89dc7c84d551e29e3335937d/tests/server.py#L788-L824) queries as part of the request), then the 1000ms jitter sleep starts, but `await_result(...)` *can* give up after another 0.9s (still 1s total but 0.1s short of the sleep finishing). This edge case happens when the floating point math satisfies [`self._reactor.seconds() > end_time`](https://github.com/element-hq/synapse/blob/963e486a70fa841f89dc7c84d551e29e3335937d/tests/server.py#L309) one iteration early so it waits exactly 1s (10 iterations). The normal case (when everything passes) is because it gets an extra [`self._reactor.advance(0.1)`](https://github.com/element-hq/synapse/blob/963e486a70fa841f89dc7c84d551e29e3335937d/tests/server.py#L312) iteration (11 iterations -> waits 1.1s).

To make it more clear, here is how we could manually drive this to completion reliably:

(`test_request_token_existing_email_inhibit_error` in `tests/rest/client/test_register.py`)
```python
channel = self.make_request(
    "POST",
    b"register/email/requestToken",
    {"client_secret": "foobar", "email": email, "send_attempt": 1},
    # We're going to drive the reactor ourselves (just below)
    await_result=False
)
# Wait for database queries to run so we can get the jitter sleep started
self.reactor.advance(0)
# Wait for any potential jitter (up to 1 second)
self.reactor.advance(Duration(seconds=1).as_secs())
```

Related to https://github.com/element-hq/synapse/pull/19394#discussion_r3157786188 and https://github.com/element-hq/synapse/pull/19734#discussion_r3157220394

I think a better fix would be to update `await_result(...)` to `self._reactor.advance(0)` before it starts its loop. This way, we don't burn any reactor advance time on things that are actually scheduled now (get rid of the foot-guns).

I'm working on improving `FakeChannel.await_result(...)` in https://github.com/element-hq/synapse/pull/19879 so I'll include the fix there ⏩ 

Overall, I think this PR should probably be reverted. We could still benefit from the comments pointing out the jitter here though.

File: tests/rest/client/test_account.py (hunk 2 above)
Comment (anoadragon453) [reply tests/rest/client/test_account.py:373]: The reasoning was that the jitter was an edge case that only applied to this endpoint, and thus these tests.

File: tests/rest/client/test_account.py (hunk 1 above)
Comment (anoadragon453) [reply tests/rest/client/test_account.py:333]: Thanks for diving in deep here! That makes sense, I didn't consider that other work could be delaying the response before the jitter even started.

I noticed that the fix was included in #19879, and the PR has now merged. Nice!

I agree that we no longer need to mess with the timeout in tests if `await_result` acts as expected, and actually it may be a bad smell to encourage tests to do so (leading to overzealous delays in unit tests). Happy to revert this PR minus the jitter comments.

> I'm not understanding the logic behind 3000ms?

It was just an arbitrarily higher value to get substantially far away from 1000ms.

File: tests/rest/client/test_account.py (hunk 2 above)
Comment (MadLittleMods) [reply tests/rest/client/test_account.py:373]: `_request_token(...)` is a custom method for the tests for this endpoint (not a general method)
